components/common/src/main/java/com/hotels/styx/common/Collections.java

```diff
@@ -0,0 +1,283 @@
+/*
+  Copyright (C) 2013-2020 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.common;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static java.util.Collections.emptyIterator;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableSet;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.Collectors.toCollection;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * {@link Collection}s cannot contain <code>null</code> elements, and will throw a {@link NullPointerException} if one is encountered.
+ *
+ * {@link Set}s, in addition, preserve ordering of source sets.
+ */
+public final class Collections {
+
+    public static <T> List<T> copyToUnmodifiableList(Iterator<? extends T> iterator) {
+        return copyToUnmodifiableList(toIterable(iterator));
+    }
+
+    public static <T> List<T> copyToUnmodifiableList(Iterable<? extends T> iterable) {
+        return unmodifiableList(stream(iterable).map(Objects::requireNonNull).collect(toList()));
+    }
+
+    public static <T> List<T> unmodifiableListOf(T... elements) {
```

Comment:
Warning: "A possible heap pollution from parametrised varargs type".

Should we allow this or not? Not sure what is the industry best practice?

Reply:
As I understand it, "heap pollution" here doesn't mean heap memory leakage, it means that you can be uncertain about the types of objects that are in the array (because at runtime this is just an `Object[]`). The example given [here](https://docs.oracle.com/javase/7/docs/technotes/guides/language/non-reifiable-varargs.html) is just within the method, not from outside, so any programming errors should be apparent and containable. I'm not sure if there is a vulnerability here that can be caused by callers to the method. I *think* not.

Reply:
Basically, you could assign `elements` to an `Object[]` reference, and then put all sorts of things in the array, and because these are all legal things to do the compiler will not warn you about them. The only chance the compiler has to warn you is your use of a parameterised array in the first place. So it does that, even if the following ode is otherwise orrect.